File: android/src/main/kotlin/com/bitmovin/player/flutter/json/JsonObjects.kt
package com.bitmovin.player.flutter.json

import com.bitmovin.analytics.api.RetryPolicy
import com.bitmovin.player.api.Player
import com.bitmovin.player.api.SeekMode
import com.bitmovin.player.api.media.MediaFilter
import com.bitmovin.player.api.media.subtitle.SubtitleTrack
import com.bitmovin.player.api.source.SourceType
import com.bitmovin.player.api.source.TimelineReferencePoint
import com.bitmovin.player.api.ui.ScalingMode
import com.bitmovin.player.casting.BitmovinCastManager
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import io.flutter.plugin.common.MethodCall
import java.security.InvalidParameterException
import kotlin.properties.ReadOnlyProperty
import kotlin.properties.ReadWriteProperty
import kotlin.reflect.KProperty

// Deserialize Json objects

private interface JStruct {
    var map: Map<*, *>

    fun toJsonString() = jacksonObjectMapper().writeValueAsString(map)
}

internal class JSource(override var map: Map<*, *>) : JStruct {
    val sourceConfig by structGetter(::JSourceConfig).require()
}

internal class JSourceConfig(override var map: Map<*, *>) : JStruct {
    val url by GetString.require()
    val type by enumGetter<SourceType>()
    val title by GetString
    val description by GetString
    val audioCodecPriority by GetStringList
    val videoCodecPriority by GetStringList
    val isPosterPersistent by GetBool
    val posterSource by GetString
    val options by structGetter(::JSourceOptions)
    val drmConfig by structGetter(::JDrmConfig)
    val analyticsSourceMetadata by structGetter(::JSourceMetadata)
    val subtitleTracks by structListGetter(::JSubtitleTrack).require()
}

internal class JSourceMetadata(override var map: Map<*, *>) : JStruct {
    val title by GetString
    val videoId by GetString
    val cdnProvider by GetString
    val path by GetString
    val isLive by GetBool
    val customData by structGetter(::JCustomData)
}

internal class JSourceOptions(override var map: Map<*, *>) : JStruct {
    val startOffset by GetDouble
    val startOffsetTimelineReference by enumGetter<TimelineReferencePoint>()
}

internal class JDrmConfig(override var map: Map<*, *>) : JStruct {
    val widevine by structGetter(::JWidevineConfig)
}

internal class JWidevineConfig(override var map: Map<*, *>) : JStruct {
    val prepareMessage by GetBool
    val prepareLicense by GetBool
    val licenseUrl by GetString
    val preferredSecurityLevel by GetString
    val shouldKeepDrmSessionsAlive by GetBool
    val httpHeaders by GetStringMap
}

internal class JStyleConfig(override var map: Map<*, *>) : JStruct {
    val isUiEnabled by GetBool
    val isHideFirstFrame by GetBool
    val supplementalPlayerUiCss by GetString
    val scalingMode by enumGetter<ScalingMode>()
    val playerUiCss by GetString
    val playerUiJs by GetString
}

internal class JPlaybackConfig(override var map: Map<*, *>) : JStruct {
    val isAutoplayEnabled by GetBool
    val isMuted by GetBool
    val isTimeShiftEnabled by GetBool
    val isTunneledPlaybackEnabled by GetBool
    val audioCodecPriority by GetStringList
    val videoCodecPriority by GetStringList
    val audioFilter by enumGetter<MediaFilter>()
    val videoFilter by enumGetter<MediaFilter>()
    val seekMode by enumGetter<SeekMode>()
}

internal class JLicensingConfig(override var map: Map<*, *>) : JStruct {
    val delay by GetInt
}

internal class JLiveConfig(override var map: Map<*, *>) : JStruct {
    val minTimeShiftBufferDepth by GetDouble
    val liveEdgeOffset by GetDouble
}

internal class JAnalyticsConfig(override var map: Map<*, *>) : JStruct {
    val licenseKey by GetString.require()
    val adTrackingDisabled by GetBool
    val randomizeUserId by GetBool
    val retryPolicy by enumGetter { stringValue ->
        when (stringValue) {
            "NoRetry" -> RetryPolicy.NO_RETRY
            "ShortTerm" -> RetryPolicy.SHORT_TERM
            "LongTerm" -> RetryPolicy.LONG_TERM
            else -> throw InvalidParameterException("Unknown enum value $stringValue")
        }
    }
    val backendUrl by GetString
    val defaultMetadata by structGetter(::JDefaultMetadata)
}

internal class JDefaultMetadata(override var map: Map<*, *>) : JStruct {
    val cdnProvider by GetString
    val customUserId by GetString
    val customData by structGetter(::JCustomData)
}

internal class JCustomData(override var map: Map<*, *>) : JStruct {
    val customData1 by GetString
    val customData2 by GetString
    val customData3 by GetString
    val customData4 by GetString
    val customData5 by GetString
    val customData6 by GetString
    val customData7 by GetString
    val customData8 by GetString
    val customData9 by GetString
    val customData10 by GetString
    val customData11 by GetString
    val customData12 by GetString
    val customData13 by GetString
    val customData14 by GetString
    val customData15 by GetString
    val customData16 by GetString
    val customData17 by GetString
    val customData18 by GetString
    val customData19 by GetString
    val customData20 by GetString
    val customData21 by GetString
    val customData22 by GetString
    val customData23 by GetString
    val customData24 by GetString
    val customData25 by GetString
    val customData26 by GetString
    val customData27 by GetString
    val customData28 by GetString
    val customData29 by GetString
    val customData30 by GetString
    val experimentName by GetString
}

internal class JPlayerConfig(override var map: Map<*, *>) : JStruct {
    val key by GetString
    val styleConfig by structGetter(::JStyleConfig)
    val playbackConfig by structGetter(::JPlaybackConfig)
    val licensingConfig by structGetter(::JLicensingConfig)
    val liveConfig by structGetter(::JLiveConfig)
    val analyticsConfig by structGetter(::JAnalyticsConfig)
    val remoteControlConfig by structGetter(::JRemoteControlConfig)
}

internal class JRemoteControlConfig(override var map: Map<*, *>) : JStruct {
    val receiverStylesheetUrl by GetString
    val customReceiverConfig by GetStringMap
    val isCastEnabled by GetBool
    val sendManifestRequestsWithCredentials by GetBool
    val sendSegmentRequestsWithCredentials by GetBool
    val sendDrmLicenseRequestsWithCredentials by GetBool
}

internal class JSubtitleTrack(override var map: Map<*, *>) : JStruct {
    constructor(subtitleTrack: SubtitleTrack) : this(emptyMap<String, Any>()) {
        url = subtitleTrack.url
        id = subtitleTrack.id
        format = subtitleTrack.mimeType
        label = subtitleTrack.label ?: subtitleTrack.id
        isDefault = subtitleTrack.isDefault
        isForced = subtitleTrack.isForced
        language = subtitleTrack.language
    }

    var url by GetSetString
    var id by GetSetString.require()
    var format by GetSetString
    var label by GetSetString.require()
    var isDefault by GetSetBool.require()
    var isForced by GetSetBool.require()
    var language by GetSetString
}

// Methods

/** Arguments for all [Player] methods. */
@JvmInline
internal value class JMethodArgs(private val call: MethodCall) {
    val asCreatePlayerArgs get() = JCreatePlayerArgs(asMap)
    val asPlayerMethodArgs get() = JPlayerMethodArg(asMap)
    val asCastManagerOptions get() = JBitmovinCastManagerOptions(asMap)
    val asCastSendMessageArgs get() = JBitmovinCastManagerSendMessageArgs(asMap)
    private val asMap get() = call.arguments as Map<*, *>
}

/** Arguments for [BitmovinCastManager.initialize] */
internal class JBitmovinCastManagerOptions(override var map: Map<*, *>) : JStruct {
    val applicationId by GetString
    val messageNamespace by GetString
}

/** Arguments for [BitmovinCastManager.sendMessage] */
internal class JBitmovinCastManagerSendMessageArgs(override var map: Map<*, *>) : JStruct {
    val message by GetString
    val messageNamespace by GetString
}

/** Arguments for [Player.create]. */
internal class JCreatePlayerArgs(override var map: Map<*, *>) : JStruct {
    val id by GetString.require()
    val playerConfig by structGetter(::JPlayerConfig).require()
}

/** Argument for all [Player] instance methods. */
internal class JPlayerMethodArg(override var map: Map<*, *>) : JStruct {
    val asDouble get() = data as Double
    val asString get() = data as String
    val asOptionalString get() = data as String?
    val asSource get() = JSource(dataAsMap)
    val asSourceConfig get() = JSourceConfig(dataAsMap)
    val asCustomData get() = JCustomData(dataAsMap)
    private val data by GetAny
    private val dataAsMap get() = data as Map<*, *>
}

internal class JPlayerViewCreateArgs(override var map: Map<*, *>) : JStruct {
    val playerId by GetString.require()
    val hasFullscreenHandler by GetBool.require()
    val isFullscreen by GetBool.require()
}

// Private property delegator

private val GetAny = castGetter<Any>()
private val GetBool = castGetter<Boolean>()
private val GetSetBool = CastDelegate<Boolean>()
private val GetInt = castGetter<Int>()
private val GetDouble = castGetter<Double>()
private val GetString = castGetter<String>()
private val GetSetString = CastDelegate<String>()
private val GetStringList = listGetter<String>()
private val GetStringMap = mapGetter<String, String>()

/** Get a field from a JSON struct */
private fun interface Getter<T> : ReadOnlyProperty<JStruct, T>

private interface GetterSetter<T> : ReadWriteProperty<JStruct, T>

/** Get a field from it's name, cast it to [I] then convert it to [O] with [build]. */
private inline fun <reified I, O> getter(crossinline build: (I) -> O) =
    Getter { thisRef, kProp ->
        (thisRef.map[kProp.name] as I?)?.let(build)
    }

private open class JsonPropertyDelegate<I, O>(val build: (I) -> O) : GetterSetter<O?> {
    override fun getValue(
        thisRef: JStruct,
        property: KProperty<*>,
    ): O? = (thisRef.map[property.name] as I?)?.let(build)

    override fun setValue(
        thisRef: JStruct,
        property: KProperty<*>,
        value: O?,
    ) {
        thisRef.map = thisRef.map.toMutableMap().apply { set(property.name, value) }
    }
}

private class JsonPropertyDelegateRequired<I, O>(
    val inner: JsonPropertyDelegate<I, O>,
) : GetterSetter<O> {
    override fun getValue(
        thisRef: JStruct,
        property: KProperty<*>,
    ): O =
        inner.getValue(thisRef, property)
            ?: throw InvalidParameterException("Missing argument ${property.name}")

    override fun setValue(
        thisRef: JStruct,
        property: KProperty<*>,
        value: O,
    ) = inner.setValue(thisRef, property, value)
}

/** Getter decorator: throw if the field is not present. */
private fun <T> Getter<T?>.require() =
    Getter<T> { thisRef, property ->
        getValue(thisRef, property)
            ?: throw InvalidParameterException("Missing argument ${property.name}")
    }

private fun <I, O> JsonPropertyDelegate<I, O>.require() = JsonPropertyDelegateRequired(this)

private inline fun <reified T> listGetter() = getter { list: List<*> -> list.map { it as T } }

private inline fun <reified E : Enum<E>> enumGetter(crossinline enumValueOf: (String) -> E) = getter(enumValueOf)

private inline fun <reified E : Enum<E>> enumGetter() = getter { name: String -> enumValueOf<E>(name, ignoreCase = true) }

private inline fun <reified E : Enum<E>> enumValueOf(
    name: String,
    ignoreCase: Boolean,
): E {
    enumValues<E>().forEach {
        if (name.equals(it.name, ignoreCase)) return it
    }
    throw InvalidParameterException("Unknown enum value $name")
}

private inline fun <reified K, reified V> mapGetter() = getter { map: Map<*, *> -> map.map { it.key as K to it.value as V }.toMap() }

private inline fun <reified T> castGetter() = getter<T, T> { it }

private class CastDelegate<T> : JsonPropertyDelegate<T, T>({ it })

private inline fun <reified T : JStruct> structGetter(crossinline build: (Map<*, *>) -> T) = getter(build)

private inline fun <reified T : JStruct> structListGetter(crossinline build: (Map<*, *>) -> T) =
    getter { list: List<Map<*, *>> -> list.map { build(it) } }

File: ios/Classes/JsonObjects.swift
// swiftlint:disable:this file_name
import BitmovinPlayer
import Foundation

internal protocol FlutterToNativeConvertible: Codable {
    associatedtype NativeObject
    func toNative() -> NativeObject
}

internal protocol NativeToFlutterConvertible {
    associatedtype FlutterObject where FlutterObject: Codable
    func toFlutter() -> FlutterObject
    func toJson() -> [String: Any]?
    func toJsonString() -> String?
}

extension NativeToFlutterConvertible {
    func toJson() -> [String: Any]? {
        guard let jsonData = try? JSONEncoder().encode(toFlutter()) else {
            return nil
        }

        return try? JSONSerialization.jsonObject(with: jsonData) as? [String: Any]
    }

    func toJsonString() -> String? {
        guard let jsonData = try? JSONEncoder().encode(toFlutter()) else {
            return nil
        }

        return String(data: jsonData, encoding: .utf8)
    }
}

internal struct FlutterAnalyticsConfig: FlutterToNativeConvertible {
    let licenseKey: String
    let retryPolicy: String
    let randomizeUserId: Bool
    let adTrackingDisabled: Bool
    let backendUrl: String
    let defaultMetadata: FlutterDefaultMetadata

    func toNative() -> AnalyticsConfig {
        let retryPolicy: RetryPolicy
        switch self.retryPolicy {
        case "ShortTerm", "NoRetry":
            retryPolicy = RetryPolicy.noRetry
        case "LongTerm":
            retryPolicy = RetryPolicy.longTerm
        default:
            retryPolicy = RetryPolicy.noRetry
        }

        return AnalyticsConfig(
            licenseKey: licenseKey,
            retryPolicy: retryPolicy,
            randomizeUserId: randomizeUserId,
            adTrackingDisabled: adTrackingDisabled,
            backendUrl: backendUrl
        )
    }
}

internal struct FlutterCustomData: FlutterToNativeConvertible {
    let customData1: String?
    let customData2: String?
    let customData3: String?
    let customData4: String?
    let customData5: String?
    let customData6: String?
    let customData7: String?
    let customData8: String?
    let customData9: String?
    let customData10: String?
    let customData11: String?
    let customData12: String?
    let customData13: String?
    let customData14: String?
    let customData15: String?
    let customData16: String?
    let customData17: String?
    let customData18: String?
    let customData19: String?
    let customData20: String?
    let customData21: String?
    let customData22: String?
    let customData23: String?
    let customData24: String?
    let customData25: String?
    let customData26: String?
    let customData27: String?
    let customData28: String?
    let customData29: String?
    let customData30: String?
    let experimentName: String?

    func toNative() -> CustomData {
        CustomData(
            customData1: customData1,
            customData2: customData2,
            customData3: customData3,
            customData4: customData4,
            customData5: customData5,
            customData6: customData6,
            customData7: customData7,
            customData8: customData8,
            customData9: customData9,
            customData10: customData10,
            customData11: customData11,
            customData12: customData12,
            customData13: customData13,
            customData14: customData14,
            customData15: customData15,
            customData16: customData16,
            customData17: customData17,
            customData18: customData18,
            customData19: customData19,
            customData20: customData20,
            customData21: customData21,
            customData22: customData22,
            customData23: customData23,
            customData24: customData24,
            customData25: customData25,
            customData26: customData26,
            customData27: customData27,
            customData28: customData28,
            customData29: customData29,
            customData30: customData30,
            experimentName: experimentName
        )
    }
}

internal struct FlutterDefaultMetadata: FlutterToNativeConvertible {
    let cdnProvider: String?
    let customUserId: String?
    let customData: FlutterCustomData

    func toNative() -> DefaultMetadata {
        DefaultMetadata(
            cdnProvider: cdnProvider,
            customUserId: customUserId,
            customData: customData.toNative()
        )
    }
}

internal struct FlutterSourceMetadata: FlutterToNativeConvertible {
    let videoId: String?
    let title: String?
    let path: String?
    let isLive: Bool?
    let cdnProvider: String?
    let customData: FlutterCustomData

    func toNative() -> SourceMetadata {
        SourceMetadata(
            videoId: videoId,
            title: title,
            path: path,
            isLive: isLive,
            cdnProvider: cdnProvider,
            customData: customData.toNative()
        )
    }
}

internal struct FlutterCueEnterEvent: Codable {
    let start: TimeInterval
    let end: TimeInterval
    let text: String?
}

extension CueEnterEvent: NativeToFlutterConvertible {
    func toFlutter() -> FlutterCueEnterEvent {
        FlutterCueEnterEvent(start: startTime, end: endTime, text: text)
    }
}

internal struct FlutterCueExitEvent: Codable {
    let start: TimeInterval
    let end: TimeInterval
    let text: String?
}

extension CueExitEvent: NativeToFlutterConvertible {
    func toFlutter() -> FlutterCueExitEvent {
        FlutterCueExitEvent(start: startTime, end: endTime, text: text)
    }
}

internal struct FlutterSubtitleTrack: FlutterToNativeConvertible {
    let url: String?
    let format: String?
    let label: String
    let id: String
    let isDefault: Bool
    let language: String?
    let isForced: Bool

    func toNative() -> SubtitleTrack {
        let subtitleTrackUrl: URL?
        if let url {
            subtitleTrackUrl = URL(string: url)
        } else {
            subtitleTrackUrl = nil
        }

        return SubtitleTrack(
            url: subtitleTrackUrl,
            format: format?.subtitleFormat ?? .webVtt,
            label: label,
            identifier: id,
            isDefaultTrack: isDefault,
            language: language,
            forced: isForced
        )
    }
}

extension SubtitleTrack: NativeToFlutterConvertible {
    func toFlutter() -> FlutterSubtitleTrack {
        FlutterSubtitleTrack(
            url: url?.absoluteString,
            format: format.stringValue,
            label: label,
            id: identifier,
            isDefault: isDefaultTrack,
            language: language,
            isForced: isForced
        )
    }
}

private extension String {
    var subtitleFormat: SubtitleFormat {
        switch self.lowercased() {
        case JsonValues.SubtitleFormat.vtt:
            return .webVtt
        case JsonValues.SubtitleFormat.ttml:
            return .ttml
        case let value where value.hasPrefix(JsonValues.SubtitleFormat.cea):
            return .cea
        default:
            return .webVtt
        }
    }
}

private extension SubtitleFormat {
    var stringValue: String {
        switch self {
        case .cea:
            return JsonValues.SubtitleFormat.cea
        case .ttml:
            return JsonValues.SubtitleFormat.ttml
        case .webVtt:
            return JsonValues.SubtitleFormat.vtt
        default:
            return JsonValues.SubtitleFormat.vtt
        }
    }
}

// Values used to handle iOS Player SDK objects that are not directly convertible
// to meaningful JSON compatible values, like for instance Obj-C enums.
private enum JsonValues {
    enum SubtitleFormat {
        static let cea = "application/cea"
        static let ttml = "application/ttml+xml"
        static let vtt = "text/vtt"
    }
}

internal struct FlutterRemoteControlConfig: FlutterToNativeConvertible {
    let receiverStylesheetUrl: String?
    let customReceiverConfig: [String: String]
    let isAirPlayEnabled: Bool
    let isCastEnabled: Bool
    let sendManifestRequestsWithCredentials: Bool
    let sendSegmentRequestsWithCredentials: Bool
    let sendDrmLicenseRequestsWithCredentials: Bool

    func toNative() -> RemoteControlConfig {
        let result = RemoteControlConfig()
        if let receiverStylesheetUrl {
            result.receiverStylesheetUrl = URL(string: receiverStylesheetUrl)
        }

        result.customReceiverConfig = customReceiverConfig
        result.isAirPlayEnabled = isAirPlayEnabled
        result.isCastEnabled = isCastEnabled
        result.sendManifestRequestsWithCredentials = sendManifestRequestsWithCredentials
        result.sendSegmentRequestsWithCredentials = sendSegmentRequestsWithCredentials
        result.sendDrmLicenseRequestsWithCredentials = sendDrmLicenseRequestsWithCredentials

        return result
    }
}

extension RemoteControlConfig: NativeToFlutterConvertible {
    func toFlutter() -> FlutterRemoteControlConfig {
        FlutterRemoteControlConfig(
            receiverStylesheetUrl: receiverStylesheetUrl?.absoluteString,
            customReceiverConfig: customReceiverConfig,
            isAirPlayEnabled: isAirPlayEnabled,
            isCastEnabled: isCastEnabled,
            sendManifestRequestsWithCredentials: sendManifestRequestsWithCredentials,
            sendSegmentRequestsWithCredentials: sendSegmentRequestsWithCredentials,
            sendDrmLicenseRequestsWithCredentials: sendDrmLicenseRequestsWithCredentials
        )
    }
}

internal struct FlutterBitmovinCastManagerOptions: FlutterToNativeConvertible {
    let applicationId: String?
    let messageNamespace: String?

    func toNative() -> BitmovinCastManagerOptions {
        let result = BitmovinCastManagerOptions()
        result.applicationId = applicationId
        result.messageNamespace = messageNamespace

        return result
    }
}

internal struct BitmovinCastManagerSendMessageArgs: Codable {
    let message: String
    let messageNamespace: String?
}

internal struct FlutterSourceConfig {
    let config: SourceConfig
    let drmMetadata: FairplayConfig.Metadata?
    let analyticsSourceMetadata: SourceMetadata?
}

internal struct FlutterSource {
    let sourceConfig: FlutterSourceConfig
    let remoteControl: SourceRemoteControlConfig?
}

internal struct SourceRemoteControlConfig {
    let castSourceConfig: FlutterSourceConfig?
}
